apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/CacheableTransformerDecorator.java

```diff
@@ -0,0 +1,194 @@
+            classloader = "00000000";
+        }
+        return classloader;
+    }
+
+    interface ClassCacheResolver {
+
+        byte[] getClassCache(ClassLoader loader, String className);
+
+        void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer);
+    }
+
+    static class MemoryCacheResolver implements ClassCacheResolver {
+        // classloaderHashcode@className -> class bytes
+        private Map<String, byte[]> classCacheMap = new ConcurrentHashMap<String, byte[]>();
+
+        @Override
+        public byte[] getClassCache(ClassLoader loader, String className) {
+            String cacheKey = getCacheKey(loader, className);
+            return classCacheMap.get(cacheKey);
+        }
+
+        @Override
+        public void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer) {
+            String cacheKey = getCacheKey(loader, className);
+            classCacheMap.put(cacheKey, classfileBuffer);
+        }
+
+        private String getCacheKey(ClassLoader loader, String className) {
+            return getClassLoaderHash(loader) + "@" + className;
+        }
+    }
+
+    static class FileCacheResolver implements ClassCacheResolver {
+
+        private final File cacheDir;
+
+        FileCacheResolver(File cacheDir) {
+            this.cacheDir = cacheDir;
+
+            //clean cache dir on exit
+            FileUtils.deleteDirectoryOnExit(cacheDir);
+        }
+
+        @Override
+        public byte[] getClassCache(ClassLoader loader, String className) {
+            // load from cache
+            File cacheFile = getCacheFile(loader, className);
+            if (cacheFile.exists()) {
+                FileInputStream fileInputStream = null;
+                try {
+                    fileInputStream = new FileInputStream(cacheFile);
+                    return IOUtils.toByteArray(fileInputStream);
+                } catch (IOException e) {
+                    logger.error("load class bytes from cache file failure", e);
+                } finally {
+                    IOUtils.closeQuietly(fileInputStream);
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer) {
+            File cacheFile = getCacheFile(loader, className);
+            FileOutputStream output = null;
+            try {
+                output = new FileOutputStream(cacheFile);
+                IOUtils.copy(new ByteArrayInputStream(classfileBuffer), output);
+            } catch (IOException e) {
+                logger.error("save class bytes to cache file failure", e);
+            } finally {
+                IOUtils.closeQuietly(output);
+            }
+        }
+
+        private File getCacheFile(ClassLoader loader, String className) {
+            String filename = getClassLoaderHash(loader) + "@" + className.replace('/', '.') + ".class";
```

Comment:
You are putting all files in one folder, NOTICE potential issue

FAT32: Maximum number of files per directory: 216 - 1 (65,535)
NTFS:Maximum number of files: 232 - 1 (4,294,967,295)
ext2/ext3:Maximum number of files per directory: ~1.3 × 1020 (performance issues past 10,000)
ext4: Maximum number of files per directory: unlimited

Material source, https://stackoverflow.com/questions/466521/how-many-files-can-i-put-in-a-directory

Reply:
I don't think this is a block, as FAT32 is rarely to see now. But just want to inform you, this could happen.

Reply:
Any suggestions?  Some dynamically generated classes may have a large number.

Reply:
Usually, file could be generated based on classloader/package....name/class name.

Do you want to change? I am sure how many people are still using FAT32.

Reply:
At the same time, I think you forgot the conflicts could happen when agents installed in several Services, but share the same folder?

Reply:
Each SkyWalking agent has a random directory like '/class-cache/class-cache-z3sdv3d4'
`File cacheDir = new File(cacheDirBase + "/class-cache-" + RandomString.make());`

Reply:
Got it. Do I need to wait for folder structure change?

Reply:
It's done, save cache class file to structure directory.